### ISSUE TITLE
Incorrect implementation of MAP

### ISSUE THREAD

**Issue by stancld** [Lightning-AI/metrics#1275]
## 🐛 Bug

After merging #1259, some discrepancies have occurred as described here https://github.com/Lightning-AI/metrics/issues/1024#issuecomment-1281424230 and #1259.

### Additional context

I'll try to investigate this more.

**Comment by Borda** [comment]
Good finding, so could we at fist place add a test to cover this correctness? :)

**Comment by aaronzs** [comment]
#1259 replaces nested for-loop with `einsum` op, while the matching process implemented in `metrics` version 0.10.0 cannot be parallelized like this. The `gt_matches` (`n_threshold x m_gt`) matrix is used as **flags** to track previous matches, so that one `gt` is matched with one `det` at a IoU threshold level. I guess the incorrectness is caused by the different behaviour when more than one `det`s can match with a `gt`.

**Comment by Borda** [comment]
@aaronzs, would you be interested in exploring this issue further? :rabbit:

**Comment by aaronzs** [comment]
@Borda Sure, I'm interested in. I'll find some test cases that can fail current implementation.

**Comment by aaronzs** [comment]
@Borda @stancld I generate random bounding boxes base on https://github.com/Lightning-AI/metrics/issues/1024#issuecomment-1281424230 and find following case that gives an incorrect result.

```python
targets = [
    {
        "boxes": torch.tensor([[4., 2., 20., 21.]]),
        "labels": torch.tensor([1]),
    }
]
preds = [
    {
        "boxes": torch.tensor(
            [
                [2., 2., 18., 13.],
                [9., 2., 21., 20.],
                [5., 4., 21., 21.],
            ]
        ).round(),
        "labels": torch.tensor([1, 1, 1]),
        "scores": torch.tensor([0.75, 0.5, 0.25]),
    }
]
```

with torchmetrics v0.6.0 (pycocotools) and torchmetrics v0.10.2

```
'map': tensor(0.2500),
'map_50': tensor(0.5000),
'map_75': tensor(0.3333),
'map_small': tensor(0.2500),
'map_medium': tensor(-1.),
'map_large': tensor(-1.),
'mar_1': tensor(0.),
'mar_10': tensor(0.6000),
'mar_100': tensor(0.6000),
'mar_small': tensor(0.6000),
'mar_medium': tensor(-1.),
'mar_large': tensor(-1.),
'map_per_class': tensor(-1.),
'mar_100_per_class': tensor(-1.)
```

with torchmetrics master branch

```
'map': tensor(0.3000),
'map_50': tensor(0.6667),
'map_75': tensor(0.3333),
'map_small': tensor(0.3000),
'map_medium': tensor(-1.),
'map_large': tensor(-1.),
'mar_1': tensor(0.),
'mar_10': tensor(0.9000),
'mar_100': tensor(0.9000),
'mar_small': tensor(0.9000),
'mar_medium': tensor(-1.),
'mar_large': tensor(-1.),
'map_per_class': tensor(-1.),
'mar_100_per_class': tensor(-1.)
```

It is a 1-target and 3-detection case, with a lower IoU threshold, e.g. 0.5, two of these detections can match with the target. `pycocotools` picks the first match while the latest implementation marks both are matched. More specifically, `det_matches` in `_evaluate_image()` is different from previous implementation.

**Comment by ananiask8** [comment]
Have you removed this metric from newer versions due to this (or other issues)? I had to install ~0.6.1 to be able to use it. I think it's gone after 0.7.1. See [https://github.com/Lightning-AI/metrics/blob/v0.6.2/torchmetrics/__init__.py#L43](v0.6.2/torchmetrics/__init__.py#L43) vs [https://github.com/Lightning-AI/metrics/blob/v0.7.1/torchmetrics/__init__.py#L43](v0.7.1/torchmetrics/__init__.py#L43).

**Comment by stancld** [comment]
> Have you removed this metric from newer versions due to this (or other issues)? I had to install ~0.6.1 to be able to use it. I think it's gone after 0.7.1. See [v0.6.2/torchmetrics/__init__.py#L43](https://github.com/Lightning-AI/metrics/blob/v0.6.2/torchmetrics/__init__.py#L43) vs [v0.7.1/torchmetrics/__init__.py#L43](https://github.com/Lightning-AI/metrics/blob/v0.7.1/torchmetrics/__init__.py#L43).

The metric was just removed from the `torchmetrics/__init__.py` due to a third-party dependency. You can still use the metric, it's just needed to be imported from `torchmetrics.detection` ;)
